### ISSUE TITLE
make pre-commit gives error on first run, then second run is fine

### ISSUE THREAD

**Issue by EgbertRijke** [UniMath/agda-unimath#524]
make pre-commit often gives an error on the first run, and then a second run is required which runs then without errors. It looks like this process could be streamlined (eg, always run twice?)

**Comment by fredrik-bakke** [comment]
That's a possibility. This morning I discovered the command `pre-commit install`, which makes `pre-commit` run automatically before every commit. This will also make your experience more streamlined.

This is added to the documentation with #514 btw.

**Comment by jonaprieto** [comment]
That behaviour is expected. In the first round, it checks the hooks and fixes any inconsistency. That's why the second round should be free of changes.

I never run: `pre-commit install`, but this is my preference; I find it annoying. I run pre-commit only as the last step and before opening a PR.

**Comment by fredrik-bakke** [comment]
Yeah, I've noticed `pre-commit` doesn't allow you to make commits when you have unstaged files for instance. This is a little annoying.

**Comment by EgbertRijke** [comment]
The behaviour wasn't initially expected to me. Perhaps we should add a sentence in the how-to's how `make pre-commit` is used, and what behaviour is expected.

**Comment by jonaprieto** [comment]
> The behaviour wasn't initially expected to me. Perhaps we should add a sentence in the how-to's how `make pre-commit` is used, and what behaviour is expected.

Sure. Btw, if you want to uninstall the hooks by default, you can run `pre-commit uninstall`. Then, you can do the next time `make pre-commit` instead.

**Comment by fredrik-bakke** [comment]
Would it be possible for the `make` target to run itself again if `pre-commit` fails only because some files were changed and none of the pre-commit hooks threw further errors? Can this be read from the error code pre-commit returns somehow? If so, then that would resolve this issue I think.

**Comment by VojtechStep** [comment]
According to [pre-commit docs](https://pre-commit.com/#command-line-interface), the exit code indicates whether the error was expected (a hook discovered an error) or unexpected (a hook blew up). I'm not sure if that's sufficient for your use case though, because I think we have hooks which can't do auto-fixes, right? In which case the most straight-forward solution I can think of is grep through the pre-commit output and check for the "files were changed by this hook" message.
On the other hand, I mostly agree with Jonathan that this is expected behavior - or rather I would expect that running `pre-commit` doesn't touch the files at all, and just does the checks; and that it then offers a flag to actually do the fixes. At least that's the workflow I'm familiar with.
